### ISSUE TITLE
`misc-const-correctness` not reported for variable storing non-const pointer result

### ISSUE THREAD

**Issue by firewave** [llvm/llvm-project#56773]
```cpp
class C
{
};

C* cb();

void f()
{
    C* c = cb();
    if (c) {
    }
}
```
https://godbolt.org/z/rvPPcE3ja

In turn the missing `const` on the pointer is detected with `WarnPointersAsValues`.

**Comment by EugeneZelenko** [comment]
@JonasToth

**Comment by JonasToth** [comment]
What is your expected code result?

```
class C
{
};

C* cb();

void f()
{
    C* c = cb();              // Original
    C* const c = cb();        // #1?
    C const* c = cb();        // #2?
    C const* const c = cb();  // #3?
    if (c) {
    }
}
```

marking the pointee `C >>const<<* ...` as `const` is not yet supported as the transformation analysis lacks that capability.

If its about the pointer in itself `C * >>const<< ...` than there is a special option to enable this. Having this on by default seemed to noisy as its not that common to mark pointers `const` (even though its useful).

**Comment by firewave** [comment]
> What is your expected code result?
> 
> ```
> class C
> {
> };
> 
> C* cb();
> 
> void f()
> {
>     C* c = cb();              // Original
>     C* const c = cb();        // #1?
>     C const* c = cb();        // #2?
>     C const* const c = cb();  // #3?
>     if (c) {
>     }
> }
> ```

Case 3 is the expected one.
 
> marking the pointee `C >>const<<* ...` as `const` is not yet supported as the transformation analysis lacks that capability.

Good to know.

> If its about the pointer in itself `C * >>const<< ...` than there is a special option to enable this. Having this on by default seemed to noisy as its not that common to mark pointers `const` (even though its useful).

That's what I meant with my comment.

**Comment by JonasToth** [comment]
With `#3` being the combination of `#1` and `#2`:

`#1` is deactivated by default as its is less common style, but can be activated. Once the options is set to `true` the warning and transformation are correct.

`#2` is a known limitation and will be addressed in future improvements of the analysis + check.

I would like to close this issue. If you do not agree, feel free to reopen it.

**Comment by firewave** [comment]
> `#2` is a known limitation and will be addressed in future improvements of the analysis + check.

> I would like to close this issue. If you do not agree, feel free to reopen it.

Since `#2` still applies I think this should stay open as an enhancement to track this.

**Comment by JonasToth** [comment]
I am in the process of creating follow up tickets for known issues that need to be addressed. I dont want to forget the issue, but its not a bug. Its better in special issue that explains it from the start and not have to scroll down to figure out that is actually a feature request.

I will link the issue.

**Comment by firewave** [comment]
> I am in the process of creating follow up tickets for known issues that need to be addressed. I dont want to forget the issue, but its not a bug. Its better in special issue that explains it from the start and not have to scroll down to figure out that is actually a feature request.
> 
> I will link the issue.

Did you create those follow-up tickets yet? Nothing has been linked yet.
